src/systems/systems/src/hydraulic/update_iterator.rs

```diff
@@ -0,0 +1,280 @@
+use crate::simulation::UpdateContext;
+use std::time::Duration;
+
+pub struct FixedStepLoop {
+    lag_time_accumulator: Duration,
+    time_step: Duration,
+    number_of_loops_this_iteration: u32,
+    number_of_loops_remaining: u32,
+}
+impl FixedStepLoop {
+    pub fn new(time_step: Duration) -> Self {
+        Self {
+            lag_time_accumulator: Duration::from_millis(0),
+            time_step,
+            number_of_loops_this_iteration: 0,
+            number_of_loops_remaining: 0,
+        }
+    }
+
+    fn time_step(&self) -> Duration {
+        self.time_step
+    }
+
+    pub fn update(&mut self, context: &UpdateContext) {
+        // Time to catch up in our simulation = new delta + time not updated last iteration
+        let time_to_catch = context.delta() + self.lag_time_accumulator;
+
+        // Number of time steps (with floating part) to do according to required time step
+        let number_of_steps_floating_point =
+            time_to_catch.as_secs_f64() / self.time_step().as_secs_f64();
+
+        if number_of_steps_floating_point < 1.0 {
+            // Can't do a full time step
+            // we can decide either do an update with smaller step or wait next iteration
+            // for now we only update lag accumulator and chose a hard fixed step: if smaller
+            // than chosen time step we do nothing and wait next iteration
+
+            // Time lag is float part only of num of steps (because is < 1.0 here) * fixed time step to get a result in time
+            self.lag_time_accumulator = Duration::from_secs_f64(
+                number_of_steps_floating_point * self.time_step().as_secs_f64(),
+            );
+            self.number_of_loops_this_iteration = 0;
+        } else {
+            // Int part is the actual number of loops to do
+            // rest of floating part goes into accumulator
+            self.number_of_loops_this_iteration = number_of_steps_floating_point.floor() as u32;
+
+            self.lag_time_accumulator = Duration::from_secs_f64(
+                (number_of_steps_floating_point - (self.number_of_loops_this_iteration as f64))
+                    * self.time_step().as_secs_f64(),
+            ); // Keep track of time left after all fixed loop are done
+        }
+
+        self.number_of_loops_remaining = self.number_of_loops_this_iteration;
+    }
+}
+impl IntoIterator for &mut FixedStepLoop {
+    type Item = Duration;
+
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut v = vec![];
+        while self.number_of_loops_remaining > 0 {
+            self.number_of_loops_remaining -= 1;
+            v.push(self.time_step);
+        }
+
+        IntoIterator::into_iter(v)
+    }
+}
+
+pub struct MaxFixedStepLoop {
```

Comment:
**question**: Reading the code in this type I've been unable to decipher what exactly it tries to do. Could you elaborate in functional terms (either here or as part of a docs comment) on what this type is trying to achieve?

Reply:
So basically you just force a max time step allowed. It can go faster, not slower.
If max step is 20ms, and sim refreshes in 10ms, it will just do one 10ms step.
If sim refreshes in 30ms, it will first do a 20ms step, then complete with the remaining 10ms step.